pkg/plugins/runtime/k8s/controllers/pod_controller.go

```diff
@@ -382,15 +426,52 @@ func (r *PodReconciler) SetupWithManager(mgr kube_ctrl.Manager, maxConcurrentRec
 		For(&kube_core.Pod{}).
 		// on Service update reconcile affected Pods (all Pods selected by this service)
 		Watches(&kube_core.Service{}, kube_handler.EnqueueRequestsFromMapFunc(ServiceToPodsMapper(r.Log, mgr.GetClient()))).
+		Watches(&kube_discovery.EndpointSlice{}, kube_handler.EnqueueRequestsFromMapFunc(EndpointSliceToPodsMapper(r.Log, mgr.GetClient()))).
 		Complete(r)
 }
 
 func ServiceToPodsMapper(l logr.Logger, client kube_client.Client) kube_handler.MapFunc {
 	l = l.WithName("service-to-pods-mapper")
 	return func(ctx context.Context, obj kube_client.Object) []kube_reconcile.Request {
+		svc := obj.(*kube_core.Service)
+		l := l.WithValues(
+			"Service",
+			kube_types.NamespacedName{Namespace: svc.Namespace, Name: svc.Name},
+		)
 		// List Pods in the same namespace as a Service
 		pods := &kube_core.PodList{}
-		if err := client.List(ctx, pods, kube_client.InNamespace(obj.GetNamespace()), kube_client.MatchingLabels(obj.(*kube_core.Service).Spec.Selector)); err != nil {
+		if len(svc.Spec.Selector) == 0 {
+			endpointSlices := &kube_discovery.EndpointSliceList{}
+			if err := client.List(
+				ctx,
+				endpointSlices,
+				kube_client.InNamespace(svc.Namespace),
+				kube_client.MatchingLabels(map[string]string{
+					kube_discovery.LabelServiceName: svc.Name,
+				}),
+			); err != nil {
+				l.Error(err, "failed to fetch EndpointSlices")
+				return nil
+			}
+			var req []kube_reconcile.Request
+			for _, slice := range endpointSlices.Items {
+				l = l.WithValues(
+					"EndpointSlice",
+					kube_types.NamespacedName{Namespace: slice.Namespace, Name: slice.Name},
+				)
+				for _, endpoint := range slice.Endpoints {
+					l = l.WithValues("Endpoint", endpoint)
```

Comment:
nitpick: we're overwriting this logger but we never use it inside these 2 loops, I guess you used it for debugging and then cleaned up the debug log, right? should we restore it on a debug level?

Reply:
I just removed it